File: Python Course/Assignment1/trie.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define SIGMA 26 // N - Size of the alphabet

typedef struct node {
    struct node *children[SIGMA + 1]; // +1 for \0
} TrieNode;

// Get index by subtracting the ASCII value of ch at the value of 'A'
int getIndex(char ch) {
    return (ch == '\0') ? 0 : ((int)ch - (int)'A' + 1);
}

// Check if node is leaf
int isLeaf(TrieNode* pNode) {
    if (pNode->children[0] == NULL)
        return 0;
    for (int i = 1; i < SIGMA + 1; i++)
        if (pNode->children[i])
            return 0;
    return 1;
}

// init new Trie node
TrieNode* init(void) {
    TrieNode *pNode = NULL;

    pNode = (TrieNode*) malloc(sizeof (TrieNode));

    if (pNode) {
        for (int i = 0; i < SIGMA + 1; i++)
            pNode->children[i] = NULL;
    }

    return pNode;
}

// Insert to Trie
void insert(TrieNode* pNode, const char* str) {
    int index;
    int len = strlen(str);

    TrieNode* pCurrNode = pNode;

    for (int i = 0; i < len; i++) {
        index = getIndex(str[i]);
        if (!pCurrNode->children[index])
            pCurrNode->children[index] = init();

        pCurrNode = pCurrNode->children[index];
    }

    if (!pCurrNode->children[0])
        pCurrNode->children[0] = init();
}

// Recursive function to remove a string from a Trie
int removeString(TrieNode *pNode, char* str) {
    int index;
    TrieNode* pNextNode;

    if (pNode == NULL || !pNode)
        return 0;

    if (*str) {
        index = getIndex(*str);
        pNextNode = pNode->children[index];
        if (!pNextNode)
            return 0;

        removeString(pNextNode, str + 1);

        if (isLeaf(pNextNode)) {
            free(pNextNode);
            pNode->children[index] = NULL;
            return 1;
        }
    }

    return 1;
}

// Search in Trie
int search(TrieNode* pNode, const char* str) {
    int index;
    int len = strlen(str);
    TrieNode* pCurrNode = pNode;

    for (int i = 0; i < len; i++) {
        index = getIndex(str[i]);

        if (!pCurrNode->children[index])
            return 0;

        pCurrNode = pCurrNode->children[index];
    }

    return (pCurrNode != NULL && pCurrNode->children[0]) ? 1 : 0;
}

int main() {
    // #define CHAR_MAX 127 // Default
    char strings[][CHAR_MAX] = { "ROMANE", "ROMANUS", "ROMULUS", "RUBENS", "RUBER", "RUB", "RUBICON", "RUBICUNDUS", "RUBY", "ROAD" }; // s - String
    char found_msg[][CHAR_MAX] = { "Not found", "Found" };

    TrieNode* trieRoot = init(); // T - Rooted tree

    int arr_size = sizeof(strings) / sizeof(strings[0]);
//    int arr_size = 10;

    for (int i = 0; i < arr_size; i++)
        insert(trieRoot, strings[i]);

    char* s = "RUBER"; // String: s

    printf("--- Before Remove ---\n");
    printf("Search %s in Trie:\t%s\n", s, found_msg[search(trieRoot, s)] );
    printf("Search %s in Trie:\t%s\n", "RUBY", found_msg[search(trieRoot, "RUBY")] );
    printf("Search %s in Trie:\t%s\n", "RUBI", found_msg[search(trieRoot, "RUBI")] );

    printf("\nDeletes %s...", s);
    if (removeString(trieRoot, s))
        printf("\n%s Successfully deleted!\n", s);


    printf("\n--- After Remove ---\n");
    printf("Search %s in Trie:\t%s\n", s, found_msg[search(trieRoot, s)] );
    printf("Search %s in Trie:\t%s\n", "RUBY", found_msg[search(trieRoot, "RUBY")] );
    printf("Search %s in Trie:\t%s\n", "RUBI", found_msg[search(trieRoot, "RUBI")] );
    return 0;
}

File: Python Course/Assignment1/trie.py
def getIndex(ch):
    return (ord(ch) - ord('A') + 1) if ord(ch) != 0 else 0


class TrieNode:
    sigma = 26  # N - Size of the alphabet

    # init node
    def __init__(self):
        self.children = [None] * (self.sigma + 1)

    # Get children by index
    def getChildren(self, index):
        return self.children[index]

    # Set children by index
    def setChildren(self, index, node):
        self.children[index] = node

    # Check if node is leaf
    def isLeaf(self):
        if self.children[0] is None:
            return False
        for i in range(1, self.sigma + 1):
            if self.children[i] is not None:
                return False
        return True


class Trie:
    # init Trie
    def __init__(self):
        self.root = TrieNode()

    # insert to Trie
    def insert(self, str):
        curr = self.root

        for ch in str:
            index = getIndex(ch)
            if curr is not None and curr.getChildren(index) is None:
                curr.setChildren(index, TrieNode())

            curr = curr.getChildren(index)

        curr.setChildren(0, TrieNode())

    # Search in Trie
    def search(self, str):
        curr = self.root

        for ch in str:
            index = getIndex(ch)
            if curr.getChildren(index) is None:
                return 0
            curr = curr.getChildren(index)
        return 1 if curr is not None and curr.getChildren(0) is not None else 0

    # Recursive function to remove a string from a Trie
    def remove(self, node, str):
        str_len = len(str)

        if node is None:
            return False

        if str_len > 0:
            index = getIndex(str[0])
            next_node = node.getChildren(index)
            if next_node is None:
                return False

            self.remove(next_node, str[1:])

            if next_node.isLeaf():
                node.setChildren(index, None)
                return False

        return True


# Test
strings = ["ROMANE", "ROMANUS", "ROMULUS", "RUBENS", "RUBER", "RUB", "RUBICON", "RUBICUNDUS", "RUBY", "ROAD"]
found_msg = ["Not found", "Found"]

root = Trie()

for s in strings:
    root.insert(s)

s = "RUBER"  # String: s

print("--- Before Remove ---")
print("Search %s in Trie:\t%s" % (s, found_msg[root.search(s)]))
print("Search %s in Trie:\t%s" % ("RUBY", found_msg[root.search("RUBY")]))
print("Search %s in Trie:\t%s" % ("RUBI", found_msg[root.search("RUBI")]))

print("\nDeletes %s..." % s)
if root.remove(root.root, s):
    print(s, "Successfully deleted!")


print("\n--- After Remove ---")
print("Search %s in Trie:\t%s" % (s, found_msg[root.search(s)]))
print("Search %s in Trie:\t%s" % ("RUBY", found_msg[root.search("RUBY")]))
print("Search %s in Trie:\t%s" % ("RUBI", found_msg[root.search("RUBI")]))
